File: lab7.java
public int findTheCity(int n, int[][] edges, int distanceThreshold) {
    // Initialize the distance matrix
    int[][] dist = new int[n][n];
    for (int i = 0; i < n; i++) {
        Arrays.fill(dist[i], Integer.MAX_VALUE);
        dist[i][i] = 0;
    }

    // Update distances based on edges
    for (int[] edge : edges) {
        int from = edge[0];
        int to = edge[1];
        int weight = edge[2];
        dist[from][to] = weight;
        dist[to][from] = weight;
    }

    // Floyd-Warshall algorithm
    for (int k = 0; k < n; k++) {
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (dist[i][k] != Integer.MAX_VALUE && dist[k][j] != Integer.MAX_VALUE) {
                    dist[i][j] = Math.min(dist[i][j], dist[i][k] + dist[k][j]);
                }
            }
        }
    }

    int minNeighbors = Integer.MAX_VALUE;
    int resultCity = -1;

    // Count reachable neighbors for each city
    for (int i = 0; i < n; i++) {
        int neighbors = 0;
        for (int j = 0; j < n; j++) {
            if (i != j && dist[i][j] <= distanceThreshold) {
                neighbors++;
            }
        }
        if (neighbors <= minNeighbors) {
            minNeighbors = neighbors;
            resultCity = i;
        }
    }

    return resultCity;
}

File: lab7.py
def findTheCity(n: int, edges: List[List[int]], distanceThreshold: int) -> int:
    # Initialize the distance matrix
    dist = [[float('inf')] * n for _ in range(n)]
    for i in range(n):
        dist[i][i] = 0

    # Update distances based on edges
    for fromi, toi, weighti in edges:
        dist[fromi][toi] = weighti
        dist[toi][fromi] = weighti

    # Floyd-Warshall algorithm
    for k in range(n):
        for i in range(n):
            for j in range(n):
                if dist[i][k] != float('inf') and dist[k][j] != float('inf'):
                    dist[i][j] = min(dist[i][j], dist[i][k] + dist[k][j])

    min_neighbors = float('inf')
    result_city = -1

    # Count reachable neighbors for each city
    for i in range(n):
        neighbors = sum(1 for j in range(n) if i != j and dist[i][j] <= distanceThreshold)
        if neighbors <= min_neighbors:
            min_neighbors = neighbors
            result_city = i

    return result_city
n = 4
edges = [[0, 1, 3], [1, 2, 1], [1, 3, 4], [2, 3, 1]]
distanceThreshold = 4
print(findTheCity(n, edges, distanceThreshold))  # Output: 3
